user/super/com/google/gwt/user/translatable/com/google/gwt/user/client/rpc/impl/ClientSerializationStreamReader.java

```diff
@@ -81,41 +83,77 @@ public void prepareToRead(String encoded) throws SerializationException {
           + "server: " + getFlags());
     }
 
-    stringTable = readJavaScriptObject();
+    stringTable = readJavaScriptObjectFromEnd();
   }
 
   public native boolean readBoolean() /*-{
-    return !!this.@com.google.gwt.user.client.rpc.impl.ClientSerializationStreamReader::results[--this.@com.google.gwt.user.client.rpc.impl.ClientSerializationStreamReader::index];
+    return !!this.@com.google.gwt.user.client.rpc.impl.ClientSerializationStreamReader::results[
+      this.@com.google.gwt.user.client.rpc.impl.AbstractSerializationStream::getVersion()() >=
```

Comment:
Before I get into "measure, _then_ cut" territory, have you profiled for any chance in perf here for large byte/int/etc array being deserialized, where instead of just an array access and a decrement, we're also reading a field? Perhaps the compiler correctly detects the general case of "nothing on the client changed the version, so this is always true"?

Reply:
No, haven't measured. Contrary to my server-side profiling environment I have no good setup as yet for client-side profiling. Do you? What I can probably try to do is an approximate timing of de-serialization of our 30MB test payload on the client side in case the browsers' built-in profilers can help me with this.

A contemporary Java VM would most certainly do what you suggest and make the additional field access an inexpensive operation. Not sure for the typical JS runtimes, though.